File: dotnet/experiments/csharp-actors/Supervisor.cs
using System.Threading.Channels;

namespace CsharpActors;

// ---------------------------------------------------------------------------
// The Supervisor is ITSELF an actor. Its mailbox receives two kinds of message:
//   * Spawn  — start a child from its factory,
//   * Exit   — a child crashed; apply the restart strategy.
//
// Strategy implemented: ONE_FOR_ONE. When one child dies, only THAT child is
// restarted; siblings are untouched (their state and loops keep running). This
// is the natural fit because each child is fully isolated behind its mailbox.
//
// RESTART INTENSITY: a child may only be restarted up to `MaxRestarts` times
// within `Window`. Exceed that and the supervisor gives up on the child and
// "escalates" (here: logs and stops restarting). This prevents a poison message
// from spinning a crash/restart loop forever.
// ---------------------------------------------------------------------------

/// <summary>A child is defined by a factory so restart = "call it again".</summary>
public sealed class ChildSpec
{
    public required string Id { get; init; }
    public required Func<IActor> Factory { get; init; }
}

// Messages understood by the supervisor's own mailbox.
public sealed record Spawn(ChildSpec Spec);

public sealed class Supervisor : IActor
{
    public string Name => "supervisor";

    // Restart-intensity budget.
    private readonly int _maxRestarts;
    private readonly TimeSpan _window;

    // Live children: id -> (running cell, its factory, recent restart timestamps).
    private readonly Dictionary<string, ChildSpec> _specs = new();
    private readonly Dictionary<string, ActorCell> _children = new();
    private readonly Dictionary<string, List<DateTime>> _restarts = new();

    // The supervisor needs a writer to ITS OWN mailbox so that children can be
    // wired to send EXIT here. Set once, right after the cell is constructed.
    private ChannelWriter<object>? _selfMailbox;

    public Supervisor(int maxRestarts = 3, TimeSpan? window = null)
    {
        _maxRestarts = maxRestarts;
        _window = window ?? TimeSpan.FromSeconds(5);
    }

    public void BindMailbox(ChannelWriter<object> selfMailbox) => _selfMailbox = selfMailbox;

    /// <summary>Look up a live child cell so the demo can send it messages.</summary>
    public ActorCell? Child(string id) => _children.TryGetValue(id, out var c) ? c : null;

    // Handled one-at-a-time, so all this bookkeeping is lock-free.
    public async ValueTask Handle(object message, IActorContext context)
    {
        switch (message)
        {
            case Spawn spawn:
                StartChild(spawn.Spec);
                break;

            case Exit exit:
                await OnChildExit(exit);
                break;
        }
    }

    private void StartChild(ChildSpec spec)
    {
        _specs[spec.Id] = spec;
        // Build a fresh IActor from the factory => fresh private state, and hand
        // the cell a writer to the supervisor's mailbox for EXIT delivery.
        var cell = new ActorCell(spec.Id, spec.Factory(), _selfMailbox);
        _children[spec.Id] = cell;
        Console.WriteLine($"[supervisor] started child '{spec.Id}'");
    }

    private async ValueTask OnChildExit(Exit exit)
    {
        Console.WriteLine($"[supervisor] EXIT received from '{exit.ChildId}': {exit.Error.Message}");

        if (!_specs.TryGetValue(exit.ChildId, out var spec))
            return; // unknown / already removed

        // Tear down the crashed cell (its loop already ended when it threw).
        if (_children.TryGetValue(exit.ChildId, out var dead))
        {
            await dead.DisposeAsync();
            _children.Remove(exit.ChildId);
        }

        // ---- restart intensity check ----
        var now = DateTime.UtcNow;
        var history = _restarts.TryGetValue(exit.ChildId, out var h) ? h : _restarts[exit.ChildId] = new();
        history.Add(now);
        history.RemoveAll(t => now - t > _window); // drop old attempts outside the window

        if (history.Count > _maxRestarts)
        {
            // Budget blown: give up on this child and escalate.
            Console.WriteLine(
                $"[supervisor] restart intensity exceeded for '{exit.ChildId}' " +
                $"({history.Count} restarts within {_window.TotalSeconds:0}s) -> ESCALATING, will not restart");
            _specs.Remove(exit.ChildId);
            return;
        }

        // ONE_FOR_ONE: restart ONLY this child. Siblings are never touched.
        Console.WriteLine($"[supervisor] one_for_one: restarting ONLY '{exit.ChildId}' " +
                          $"(restart #{history.Count} within window)");
        StartChild(spec);
    }
}

File: dotnet/experiments/fsharp-actors/Supervisor.fs
namespace FSharpActors

open System

// -----------------------------------------------------------------------------
// Supervisor actor: itself a MailboxProcessor, holding a map of live children
// and their factories. Implements a ONE_FOR_ONE restart strategy with a restart
// intensity budget (max 3 restarts within 5s => escalate / give up).
//
// F#-vs-C# HIGHLIGHT #3 (immutable supervisor state + selective concerns):
//   The supervisor's own state (which children are alive, recent restart times)
//   is threaded immutably through its loop exactly like the worker's counter.
//   No shared mutable dictionary, no lock: mail arrives one at a time, so
//   ChildExit is handled to completion before the next message. A child crash is
//   just another case in the same exhaustive `match`.
// -----------------------------------------------------------------------------

module Supervisor =

    /// How a supervisor knows to (re)build a child: its id + a factory that,
    /// given the supervisor mailbox, produces a brand-new worker.
    type ChildSpec =
        { Id: int
          Factory: int -> MailboxProcessor<SupervisorMsg> -> MailboxProcessor<WorkerMsg> }

    // Restart intensity budget.
    let private maxRestarts = 3
    let private withinWindow = TimeSpan.FromSeconds 5.0

    /// Start a supervisor for the given child specs. Returns the supervisor
    /// mailbox; callers interact with children by posting `Route(id, msg)`.
    let start (specs: ChildSpec list) : MailboxProcessor<SupervisorMsg> =
        MailboxProcessor.Start(fun inbox ->

            // Spawn the initial children. Each factory captures `inbox` (this
            // supervisor) so children can report crashes back to us.
            let spawnAll () =
                specs
                |> List.map (fun spec -> spec.Id, spec.Factory spec.Id inbox)
                |> Map.ofList

            let specById =
                specs |> List.map (fun s -> s.Id, s) |> Map.ofList

            // Supervisor state is immutable:
            //   children     : id -> current live worker mailbox
            //   restartTimes : recent restart timestamps (for the intensity budget)
            let rec loop (children: Map<int, MailboxProcessor<WorkerMsg>>)
                         (restartTimes: DateTime list) =
                async {
                    let! msg = inbox.Receive()

                    match msg with
                    | Route (id, workerMsg) ->
                        match Map.tryFind id children with
                        | Some child -> child.Post workerMsg
                        | None -> Log.line (sprintf "[supervisor] no live child %d to route to" id)
                        return! loop children restartTimes

                    | ChildExit (id, error) ->
                        Log.line (sprintf "[supervisor] EXIT from child %d: %s" id error.Message)

                        // Prune restart history to just the last `withinWindow`.
                        let now = DateTime.UtcNow
                        let recent =
                            restartTimes
                            |> List.filter (fun t -> now - t < withinWindow)

                        if List.length recent >= maxRestarts then
                            // Intensity budget blown: ESCALATE / give up. We stop
                            // the offending child from being resurrected. (In a
                            // full OTP tree this would terminate the supervisor and
                            // propagate to *its* supervisor.)
                            Log.line (sprintf "[supervisor] restart intensity exceeded (%d in %gs) -> GIVING UP on child %d"
                                          maxRestarts withinWindow.TotalSeconds id)
                            let children' = Map.remove id children
                            return! loop children' recent
                        else
                            // ONE_FOR_ONE: restart ONLY the failed child. All other
                            // children keep running with their state intact.
                            match Map.tryFind id specById with
                            | Some spec ->
                                Log.line (sprintf "[supervisor] restarting child %d (one_for_one) -> fresh state" id)
                                let fresh = spec.Factory spec.Id inbox
                                let children' = Map.add id fresh children
                                return! loop children' (now :: recent)
                            | None ->
                                return! loop (Map.remove id children) recent
                }

            loop (spawnAll ()) [])
